### ISSUE TITLE
New filter : only_tty ?

### ISSUE THREAD

**Issue by mackermann2** [a2o/snoopy#83]
Hello, 

Is there possible to add a new filter "only_tty" to exclude all lines which contains "tty:(none)" like this : 

[uid:0 sid:15400 tty:(none) cwd:/product/scripts filename:/usr/bin/awk]: awk -F , {printf "$3"_"$4 "\n"} /var/www/html/test/csv/20150709.csv
[uid:0 sid:25283 tty:(none) cwd:/product/scripts filename:/bin/ps]: /bin/ps auxw
[uid:0 sid:25283 tty:(none) cwd:/product/scripts filename:/bin/grep]: /bin/grep -v grep
[uid:0 sid:32716 tty:(none) cwd:/root filename:/bin/date]: /bin/date +[%x %H:%M:%S]

I would like to log only commands in a shell user. 

Best regards,
Matthieu

**Comment by bostjan** [comment]
Hi Matthieu,

it is possible :)

Will you give it a shot and submit a pull request?
I believe it is acually not that hard. Check doc/CONTRIBUTING for basic info for new filter development, and src/filter/* for examples.

b.

**Comment by bostjan** [comment]
Also, code for checking if command was executed on tty is already present in src/datasource/tty* files.

b.

**Comment by mackermann2** [comment]
Hi Bostjan, 

Thanks for your response. 
Actually i'm not a guru guy in C language but I will try to find a solution myself.
I thinks it's possible :)

Matthieu

**Comment by bostjan** [comment]
Alright, I like the spirit.

If you have questions, do not hesitate to ask. But majority of work should be copy-pasting.

b.

**Comment by urq** [comment]
Hi there, I had an urgent need for this only_tty filter, and after implementing it reduced noise in my server auth logs by 99%.  I thought I'd drop the pull request here, though there is one omission--it doesn't include tests yet (though it is currently passing tests in the wild ;) ).

**Comment by bostjan** [comment]
Excellent!

It looks good, I will merge it next week. I have only 1 non-complaint: in alphabet, letter t comes before u and not after :)

About tests: do not worry, as this one is a bit tricky to test. I will create it.

Did you by any chance run it by valgrind? (configure --enable-everything; make valgrind)

b.

**Comment by urq** [comment]
I identified a copypasta issue which I'll fix, but while I'm in there I can alphabetize :)

When I tried running it by valgrind, my `configure` left this warning on the end:

```=== configuring in lib/iniparser (/home/blah/snoopy/lib/iniparser)
configure: WARNING: no configuration information is in lib/iniparser```

and then my `make valgrind` blew up with errors looking in that directory, like so:

```blah@blah:~/snoopy$ make valgrind
make  all-recursive
make[1]: Entering directory `/home/blah/snoopy'
Making all in etc
make[2]: Entering directory `/home/blah/snoopy/etc'
make[2]: Nothing to be done for `all'.
make[2]: Leaving directory `/home/blah/snoopy/etc'
Making all in lib
make[2]: Entering directory `/home/blah/snoopy/lib'
Making all in iniparser
make[3]: Entering directory `/home/blah/snoopy/lib/iniparser'
make[3]: *** No rule to make target `all'.  Stop.
make[3]: Leaving directory `/home/blah/snoopy/lib/iniparser'
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory `/home/blah/snoopy/lib'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/home/blah/snoopy'
make: *** [all] Error 2```

I'm guessing I'm missing something, but I'm not sure what... note, these errors go away when I configure with `--disable-config-file`.

**Comment by bostjan** [comment]
I think you did not initialize git submodule for iniparser. git submodule init & git submodule update should do it.

b.

**Comment by urq** [comment]
Ah, I did forget that, but even after running `git submodule init && git submodule update`, I'm getting the same.  Looking in the `./configure` file, it looks like it's looking for a `configure` or `configure.gnu` or `configure.in` file in `lib/iniparser`, but isn't finding any of those.  This is what is in `lib/iniparser`:
```
blah@blah:~/snoopy/lib/iniparser$ ls
AUTHORS  build  config.h.in  configure.ac  configure.scan  doc  example  html  INSTALL  LICENSE  Makefile.am  README.md  src  test
```

Any thoughts?  I'm a bit out of my depth with the C toolchain, I'm afraid.

**Comment by bostjan** [comment]
Yeah, it was missing from doc/INSTALL.

Also, please re-run ./bootstrap.sh which does all the necessary things before ./configure step.

**Comment by bostjan** [comment]
Updated doc about this too :)

**Comment by urq** [comment]
Got it.  Thanks for the debugging help.  Valgrind looks good.  No leaks, and the open fds look like they're coming from elsewhere.
```
All done.
==18987==
==18987== FILE DESCRIPTORS: 5 open at exit.
==18987== Open file descriptor 4: /var/lib/sss/mc/passwd
==18987==    at 0x534B620: __open_nocancel (syscall-template.S:81)
==18987==    by 0x5E39FB1: ???
==18987==    by 0x5E39BE3: ???
==18987==    by 0x5E3A3D8: ???
==18987==    by 0x5E38085: ???
==18987==    by 0x53202EB: getpwuid_r@@GLIBC_2.2.5 (getXXbyYY_r.c:266)
==18987==    by 0x405DCA: snoopy_datasource_eusername (eusername.c:72)
==18987==    by 0x4076D5: snoopy_debug_test_all_datasources (libsnoopy-debug-addons.c:88)
==18987==    by 0x401C21: main (snoopy-run-everything.c:74)
==18987==
==18987== Open file descriptor 3: /var/lib/sss/mc/group
==18987==    at 0x534B620: __open_nocancel (syscall-template.S:81)
==18987==    by 0x5E39FB1: ???
==18987==    by 0x5E39BE3: ???
==18987==    by 0x5E3A7D8: ???
==18987==    by 0x5E38A55: ???
==18987==    by 0x531EB0B: getgrgid_r@@GLIBC_2.2.5 (getXXbyYY_r.c:266)
==18987==    by 0x405B2A: snoopy_datasource_egroup (egroup.c:72)
==18987==    by 0x4076D5: snoopy_debug_test_all_datasources (libsnoopy-debug-addons.c:88)
==18987==    by 0x401C21: main (snoopy-run-everything.c:74)
==18987==
==18987== Open file descriptor 2: /dev/pts/1
==18987==    <inherited from parent>
==18987==
==18987== Open file descriptor 1: /dev/pts/1
==18987==    <inherited from parent>
==18987==
==18987== Open file descriptor 0: /dev/pts/1
==18987==    <inherited from parent>
==18987==
==18987==
==18987== HEAP SUMMARY:
==18987==     in use at exit: 0 bytes in 0 blocks
==18987==   total heap usage: 141 allocs, 141 frees, 118,744 bytes allocated
==18987==
==18987== All heap blocks were freed -- no leaks are possible
==18987==
==18987== For counts of detected and suppressed errors, rerun with: -v
==18987== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by bostjan** [comment]
Yeah, looks good. Could you please paste this output into separate issue, so I will look into these open file descriptors next week. Tnx.

b.

**Comment by urq** [comment]
Will do.  Also fixed copypasta error and alphabetization.  Thanks again.

**Comment by bostjan** [comment]
Merged and released in v2.4.4

**Comment by mackermann2** [comment]
Hi Bostjan,

Thanks a lot for this new release. But it seems to not working. Indeed : 

- my configuration file : 
egrep -v "^;|^$" /etc/snoopy.ini
[snoopy]
message_format = "[%{datetime}] workdir=%{cwd} login=%{login} pid=%{pid} tty=%{tty} tty_username=%{tty_username} cmdline=%{cmdline}"
filter_chain = "only_tty"
output = file:/var/log/snoopy.log

- my crontab : 
crontab -l
*/1 * * * * ls /tmp 2>&1

- my log file : 

[2015-08-28T12:13:20+0200] workdir=/root login=root pid=19349 tty=/dev/pts/0 tty_username=root cmdline=vim /etc/snoopy.ini
[2015-08-28T12:14:01+0200] workdir=/root login=root pid=19369 tty=ERROR(ttyname_r->EUNKNOWN) tty_username=ERROR(ttyname_r->EUNKNOWN) cmdline=ls /tmp
[2015-08-28T12:15:01+0200] workdir=/root login=root pid=19409 tty=ERROR(ttyname_r->EUNKNOWN) tty_username=ERROR(ttyname_r->EUNKNOWN) cmdline=ls /tmp
[2015-08-28T12:16:01+0200] workdir=/root login=root pid=19448 tty=ERROR(ttyname_r->EUNKNOWN) tty_username=ERROR(ttyname_r->EUNKNOWN) cmdline=ls /tmp
[2015-08-28T12:17:01+0200] workdir=/root login=root pid=19486 tty=ERROR(ttyname_r->EUNKNOWN) tty_username=ERROR(ttyname_r->EUNKNOWN) cmdline=ls /tmp

Commands not executed in a tty are logged (tty=ERROR(ttyname_r->EUNKNOWN))

- Others background commands like "sa" are logged too : 

[2015-08-28T12:30:01+0200] workdir=/usr/lib64/sa login=root pid=20024 tty=ERROR(ttyname_r->EUNKNOWN) tty_username=ERROR(ttyname_r->EUNKNOWN) cmdline=/usr/lib64/sa/sadc -F -L -S DISK 60 10 -
[2015-08-28T12:30:21+0200] workdir=/root login=root pid=20040 tty=/dev/pts/1 tty_username=root cmdline=grep tty=ERROR /var/log/snoopy.log
[2015-08-28T12:30:21+0200] workdir=/root login=root pid=20041 tty=ERROR(ttyname_r->EUNKNOWN) tty_username=ERROR(ttyname_r->EUNKNOWN) cmdline=uniq



Any ideas ? 

Matthieu

**Comment by bostjan** [comment]
It is true I only created test for this new filter, which passed once debugged, on all systems. I just tested this manually, with your configuration, and it works with no hassle (ubuntu 14.04 lxc container).

What OS are you using?

**Comment by bostjan** [comment]
Can you apply this diff and paste the results that are logged?
https://gist.github.com/bostjan/7a944aaf41701d724c81

It would log the return value of ttyname_r() at the end of error message.

Desc:
In the code of datasource tty, I covered all return values listed in manpage for ttyname_r, so this is something new. It didn't work with 2.4.3 too, right?

b.

**Comment by mackermann2** [comment]
Hello. I've made an RPM package for RHEL 6.5:

cat /etc/redhat-release
Red Hat Enterprise Linux Server release 6.5 (Santiago)

The specification file : 

Name: snoopy
Version: 2.4.4
Release: 1
Summary: Snoopy is a wrapper around execve() that logs all executed commands by all users and processes to syslog.
Vendor: SFR CAV800
Group: Application/Monitoring
License: GPLv2
BuildRoot: %{_tmppath}/snoopy
BuildArch: x86_64
#BuildRequires:
#Requires:

%define _rpmfilename snoopy-%{version}.rpm

%description
Snoopy is designed to aid a sysadmin by providing a log of commands executed. Snoopy is completely transparent to the user and applications. It is linked into programs to provide a wrapper around calls to execve(). Logging is done via syslog.

%prep
#%setup -q
rm -rf %{_topdir}/BUILD/snoopy-2.4.4
tar zxvf %{_topdir}/SOURCES/snoopy-2.4.4.tar.gz -C %{_topdir}/BUILD/

%pre

%build
cd %{_topdir}/BUILD/snoopy-2.4.4
%configure --enable-config-file --enable-filtering
make DESTDIR=%{buildroot}

%install
rm -rf %{buildroot}
cd %{_topdir}/BUILD/snoopy-2.4.4
make install DESTDIR=%{buildroot}
make enable

%clean
rm -rf %{buildroot}
rm -rf %{_topdir}/BUILD/*

%files
%defattr(-,root,root,-)
%doc
/etc/*
/usr/*

%post -p /usr/sbin/snoopy-enable
echo Install successfull.

%preun -p /usr/sbin/snoopy-disable

%changelog
* Tue Nov 4 2014 Jeremy Brown <jwbrown77@yahoo.com> 2.0.0rc12-1
- Initial RPM spec for snoopy.

**Comment by mackermann2** [comment]
Modification of "tty.c" has generated an error during "make" : 

tty.c: In function ‘snoopy_datasource_tty’:
tty.c:68: error: ‘re’ undeclared (first use in this function)
tty.c:68: error: (Each undeclared identifier is reported only once
tty.c:68: error: for each function it appears in.)
tty.c:69: error: expected ‘)’ before ‘}’ token
tty.c:69: error: expected ‘;’ before ‘}’ token

In your URL https://gist.githubusercontent.com/bostjan/7a944aaf41701d724c81/raw/1bf742f1addaabda50fee3d82f61f01279cd349a/ttyname_r-return-detection.diff, end of line seems incorrect ( no ; character ).

Is this normal ?

**Comment by bostjan** [comment]
Nope, something has cut the end of line. Fixed, pls try again.

**Comment by mackermann2** [comment]
The result is the following one (return code 22): 

[2015-08-28T15:20:01+0200] workdir=/root login=root pid=27068 tty=ERROR(ttyname_r->EUNKNOWN:22) tty_username=ERROR(ttyname_r->EUNKNOWN) cmdline=ls /tmp
[2015-08-28T15:20:01+0200] workdir=/root login=root pid=27069 tty=ERROR(ttyname_r->EUNKNOWN:22) tty_username=ERROR(ttyname_r->EUNKNOWN) cmdline=/usr/lib64/sa/sa1 60 10
[2015-08-28T15:20:01+0200] workdir=/usr/lib64/sa login=root pid=27069 tty=ERROR(ttyname_r->EUNKNOWN:22) tty_username=ERROR(ttyname_r->EUNKNOWN) cmdline=/usr/lib64/sa/sadc -F -L -S DISK 60 10 -

**Comment by bostjan** [comment]
BTW: Tried it in centos 6 LXC container, works as expected.

**Comment by bostjan** [comment]
This seems to be the issue:
https://sourceware.org/bugzilla/show_bug.cgi?id=12340

Please tell me your glibc version? This should be resolved in 2.16.

I tested it on glibc 2.7 too, it works as expected there too.

**Comment by bostjan** [comment]
There seem to be two issues at play:
1. the one mentioned above - can you do yum upgrade and retest with latest glibc?
2. If #1 is fixed, this EINVAL will change to ERANGE, which is good and predicted, but this still does not explain why buffer of 4096 bytes is too small for your path to TTY. This is weird.

Can you play around with buffer lengths in datasource/tty.h and report, if tty=ERROR(ttyname_r->EUNKNOWN) changes to something that can actually resembles path to tty?

**Comment by mackermann2** [comment]
My version is an old version : "glibc-2.12-1.47.el6.x86_64".
I'm going to try to update (yum upgrade doesnt work. It's already the last version in my repository).

I will let you know.

**Comment by bostjan** [comment]
I checked one RHEL I manage, set up in mid 2013, now it at .166 glibc version, yum upgraded up to rhel 6.7.

**Comment by bostjan** [comment]
Closing, no feedback.
